torchtext/prototype/datasets/text_classification.py

```diff
@@ -0,0 +1,613 @@
+    for cls, tokens in iterator:
+        yield cls, iter(map(lambda x: vocab[x],
+                        filter(lambda x: x not in removed_tokens, tokens)))
+
+
+def _imdb_iterator(key, extracted_files, tokenizer, ngrams, yield_cls=False):
+    for fname in extracted_files:
+        if 'urls' in fname:
+            continue
+        elif key in fname and ('pos' in fname or 'neg' in fname):
+            with io.open(fname, encoding="utf8") as f:
+                label = 1 if 'pos' in fname else 0
+                if yield_cls:
+                    yield label, ngrams_iterator(tokenizer(f.read()), ngrams)
+                else:
+                    yield ngrams_iterator(tokenizer(f.read()), ngrams)
+
+
+class TextClassificationDataset(torch.utils.data.Dataset):
+    """Defines an abstract text classification datasets.
+       Currently, we only support the following datasets:
+
+             - AG_NEWS
+             - SogouNews
+             - DBpedia
+             - YelpReviewPolarity
+             - YelpReviewFull
+             - YahooAnswers
+             - AmazonReviewPolarity
+             - AmazonReviewFull
+
+    """
+
+    def __init__(self, vocab, data, labels):
+        """Initiate text-classification dataset.
+
+        Arguments:
+            vocab: Vocabulary object used for dataset.
+            data: a list of label/tokens tuple. tokens are a tensor after
+                numericalizing the string tokens. label is an integer.
+                [(label1, tokens1), (label2, tokens2), (label2, tokens3)]
+            label: a set of the labels.
+                {label1, label2}
+
+        Examples:
+            See the examples in examples/text_classification/
+
+        """
+
+        super(TextClassificationDataset, self).__init__()
+        self._data = data
+        self._labels = labels
+        self._vocab = vocab
+
+    def __getitem__(self, i):
+        return self._data[i]
+
+    def __len__(self):
+        return len(self._data)
+
+    def __iter__(self):
+        for x in self._data:
+            yield x
+
+    def get_labels(self):
+        return self._labels
+
+    def get_vocab(self):
+        return self._vocab
+
+
+def _generate_data_iterators(dataset_name, root, ngrams, tokenizer, data_select):
+    if not tokenizer:
+        tokenizer = get_tokenizer("basic_english")
+
+    if not set(data_select).issubset(set(('train', 'test'))):
+        raise TypeError('Given data selection {} is not supported!'.format(data_select))
+
+    dataset_tar = download_from_url(URLS[dataset_name], root=root)
+    extracted_files = extract_archive(dataset_tar)
+
+    path = {}
+    if dataset_name != 'IMDB':
+        for fname in extracted_files:
+            if fname.endswith('train.csv') and 'train' in data_select:
+                path['train'] = fname
+            if fname.endswith('test.csv') and 'test' in data_select:
+                path['test'] = fname
+
+    iters_group = {}
+    if 'train' in data_select:
```

Comment:
Both "_imdb_iterator" and "_csv_iterator" have a similar interface, except that _csv_iterator is passed a path. Instead of accepting a `path['train']` it could simply accept `'train'` like _imdb_iterator does. Then you only need to create an iterator via something like `_gen_imdb_iterator` and all of this branching etc. can disappear.